hail/python/hail/table.py

```diff
@@ -1202,6 +1202,44 @@ def write(self, output: str, overwrite = False, stage_locally: bool = False,
         Env.backend().execute(TableWrite(self._tir, output, overwrite, stage_locally, _codec_spec))
 
     def _show(self, n, width, truncate, types):
+        return Table._Show(self, n, width, truncate, types)
+
+    class _Show:
+        def __init__(self, table, n, width, truncate, types):
+            self.table = table
+            self.n = n
+            self.width = width
+            self.truncate = truncate
+            self.types = types
+
+        def __str__(self):
+            return self.table._ascii_str(self.n, self.width, self.truncate, self.types)
+
+        def __repr__(self):
+            return self.__str__()
+
+        def _repr_html_(self):
+            return self.table._html_str(self.n, self.types)
+
+    @staticmethod
+    def _hl_repr(v):
+        if v.dtype == hl.tfloat32 or v.dtype == hl.tfloat64:
+            s = hl.format('%.2e', v)
+        elif isinstance(v.dtype, hl.tarray):
+            s = "[" + hl.delimit(hl.map(Table._hl_repr, v), ",") + "]"
+        elif isinstance(v.dtype, hl.tset):
+            s = "{" + hl.delimit(hl.map(Table._hl_repr, hl.array(v)), ",") + "}"
+        elif isinstance(v.dtype, hl.tdict):
+            s = "{" + hl.delimit(hl.map(lambda x: Table._hl_repr(x[0]) + ":" + Table._hl_repr(x[1]), hl.array(v)), ",") + "}"
+        elif v.dtype == hl.tstr:
+            s = hl.str('"') + hl.expr.functions._escape_string(v) + '"'
+        elif isinstance(v.dtype, (hl.tstruct, hl.ttuple)):
+            s = "(" + hl.delimit([Table._hl_repr(v[i]) for i in range(len(v))], ",") + ")"
```

Comment:
needs to check for empty array - that can't be type-imputed

Reply:
(this is your test failure)

Reply:
le sigh. fixed.